server/src/internalClusterTest/java/org/opensearch/action/bulk/AppendOnlyIndicesIT.java

```diff
@@ -0,0 +1,170 @@
+            nodeStats.getNodes().stream().filter((s) -> s.getNode().isDataNode()).collect(Collectors.toList())
+        );
+        assertAcked(
+            client().admin()
+                .indices()
+                .prepareCreate("index")
+                .setSettings(
+                    Settings.builder()
+                        .put("index.number_of_replicas", 0)
+                        .put("index.number_of_shards", 1)
+                        .put("index.append_only.enabled", true)
+                )
+        );
+        ensureGreen("index");
+        logger.info("unlucky node: {}", unluckyNode.getNode());
+        // create a transport service that throws a ConnectTransportException for one bulk request and therefore triggers a retry.
+        for (NodeStats dataNode : nodeStats.getNodes()) {
+            if (exceptionThrown.get()) {
+                break;
+            }
+
+            MockTransportService mockTransportService = ((MockTransportService) internalCluster().getInstance(
+                TransportService.class,
+                dataNode.getNode().getName()
+            ));
+            mockTransportService.addSendBehavior(
+                internalCluster().getInstance(TransportService.class, unluckyNode.getNode().getName()),
+                (connection, requestId, action, request, options) -> {
+                    connection.sendRequest(requestId, action, request, options);
+                    if (action.equals(TransportShardBulkAction.ACTION_NAME) && exceptionThrown.compareAndSet(false, true)) {
+                        logger.debug("Throw ConnectTransportException");
+                        throw new ConnectTransportException(connection.getNode(), action);
+                    }
+                }
+            );
+        }
+
+        BulkRequestBuilder bulkBuilder = client.prepareBulk();
+
+        for (int i = 0; i < numDocs; i++) {
+            XContentBuilder doc = null;
+            doc = jsonBuilder().startObject().field("foo", "bar").endObject();
+            bulkBuilder.add(client.prepareIndex("index").setSource(doc));
+        }
+
+        BulkResponse response = bulkBuilder.get();
+        for (BulkItemResponse singleIndexResponse : response.getItems()) {
+            // Retry will not create a new version.
+            assertThat(singleIndexResponse.getVersion(), equalTo(1L));
+        }
+
+        response = client().prepareBulk().add(client().prepareUpdate("index", "0").setDoc("foo", "updated")).get();
+        assertThat(
+            response.getItems()[0].getFailureMessage(),
+            containsString("Operation [UPDATE] is not allowed for append only index index;")
+        );
+
+        response = client().prepareBulk().add(client().prepareDelete("index", "0")).get();
+        assertThat(
+            response.getItems()[0].getFailureMessage(),
+            containsString("Operation [DELETE] is not allowed for append only index index;")
+        );
+    }
```

Comment:
Why are we testing update/delete in retry test?

Reply:
These calls will be blocked at the transport layer itself. I was thinking around covering different scenarios. Will remove this.